XIPs/xip-35-message-sender-signature.md

```diff
@@ -0,0 +1,60 @@
+---
+xip: 35
+title: Message sender identifier in topic
+description: A system for detecting who sent a message in a topic from the Notification Server
+author: Noé Malzieu
+status: Draft
+type: Standards Track
+category: XRC
+created: 2023-12-06
+---
+
+
+## Abstract
+
+This XRC proposes to add a signature to public envelopes for messages, that a notification server could use to detect if one of the subscribed users has sent the message and not send him the notification (i.e. avoid receiving notifications for your own messages).
+
+## Motivation
+
+Right now, a message sent in a topic needs to have its payload decoded to understand who sent the message.
+When a notification server sees a message in a topic and decides to sent it to one of its subscribers, it has no way to know if the message is from the subscriber or not.
+This means a mobile XMTP client will subscribe to a topic then receive notifications for its own messages in the topic.
+This would be fine if the XMTP client could just drop the notification after having decoded the payload.
+On Apple devices, to be able to decode the payload before showing the notification, we must use a [Notification Service Extension](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension) - and the only way to drop a notification from a Notification Service Extension is to obtain a specific entitlement from Apple: [com.apple.developer.usernotifications.filtering](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering)
+Obtaining this entitlement from Apple has proven very hard - they might consider XMTP clients too web3, and they just don't provide this entitlement to apps that provide financial services.
+Specifically, Coinbase Wallet never managed to obtain this entitlement, and Converse had it for a few months before losing it.
+
+## Specification
+
+The SDKs would generate a private/public key pair for each topic.
+The private/public key pair should always be the same for a given user and a given topic - not dependent on installations / SDK language.
```

Comment:
My suggestion would be that each user would have a signing key that would be synced between their devices for v1/v2 and be per-installation and not synced for group chats/v3. We would generate a per-topic key by doing `HKDF(root_key, salt=$topic)`.

Reply:
Not sure I get the difference between an "installation" and a "device"?
Goal being that I generate the same per-topic key on Coinbase Wallet & Converse

Reply:
> Not sure I get the difference between an "installation" and a "device"?

"Installation" is a little more specific than "device", because in the XMTP case there can be multiple different apps (Coinbase Wallet, Converse) installed on the same device. We sometimes forget and use the word "device" interchangeably, because a lot of the literature uses that term.

> Goal being that I generate the same per-topic key on Coinbase Wallet & Converse

@neekolas's suggestion would work for this - the signing key is synced to all installations, and all installations perform the same `HKDF` to get the same per-topic key.

Reply:
> We would generate a per-topic key by doing HKDF(root_key, salt=$topic).

I'm not sure I understand this proposal for v3. The per-topic key needs to be the same among all installations, so what is "root_key" referring to here?